tests/tls_e2e/common/utility.cpp

```diff
@@ -0,0 +1,318 @@
+    mbedtls_rsa_context* rsa_ctx = NULL;
+    uint8_t* modulus = NULL;
+    size_t modulus_size = 0;
+    int res = 0;
+    bool ret = false;
+    unsigned char* signer = NULL;
+
+    signer = (unsigned char*)malloc(signer_id_buf_size);
+    if (signer == NULL)
+    {
+        OE_TRACE_ERROR("Out of memory\n");
+        goto done;
+    }
+
+    OE_TRACE_INFO("Verify connecting client's identity\n");
+    OE_TRACE_INFO(
+        "public key buffer size[%lu]\n", sizeof(siging_public_key_buf));
+    OE_TRACE_INFO("public key\n[%s]\n", siging_public_key_buf);
+
+    mbedtls_pk_init(&ctx);
+    res = mbedtls_pk_parse_public_key(
+        &ctx,
+        (const unsigned char*)siging_public_key_buf,
+        siging_public_key_buf_size);
+    if (res != 0)
+    {
+        OE_TRACE_ERROR("mbedtls_pk_parse_public_key failed with %d\n", res);
+        goto done;
+    }
+    OE_TRACE_INFO(
+        "siging_public_key_buf_size=%ld\n", siging_public_key_buf_size);
+
+    pk_type = mbedtls_pk_get_type(&ctx);
+    if (pk_type != MBEDTLS_PK_RSA)
+    {
+        OE_TRACE_ERROR("mbedtls_pk_get_type had incorrect type: %d\n", res);
+        goto done;
+    }
+    OE_TRACE_INFO("This public sigining key is a rsa key \n");
+
+    rsa_ctx = mbedtls_pk_rsa(ctx);
+    modulus_size = mbedtls_rsa_get_len(rsa_ctx);
+    OE_TRACE_INFO("modulus_size = [%zu]\n", modulus_size);
+    modulus = (uint8_t*)malloc(modulus_size);
+    if (modulus == NULL)
+    {
+        OE_TRACE_ERROR(
+            "malloc for modulus failed with size %zu:\n", modulus_size);
+        goto done;
+    }
+
+    res = mbedtls_rsa_export_raw(
+        rsa_ctx, modulus, modulus_size, NULL, 0, NULL, 0, NULL, 0, NULL, 0);
+    if (res != 0)
+    {
+        OE_TRACE_ERROR("mbedtls_rsa_export failed with %d\n", res);
+        goto done;
+    }
+
+    // Reverse the modulus and compute sha256 on it.
+    for (size_t i = 0; i < modulus_size / 2; i++)
+    {
+        uint8_t tmp = modulus[i];
+        modulus[i] = modulus[modulus_size - 1 - i];
+        modulus[modulus_size - 1 - i] = tmp;
+    }
+
+    // Calculate the MRSIGNER value which is the SHA256 hash of the
+    // little endian representation of the public key modulus. This value
+    // is populated by the signer_id sub-field of a parsed oe_report_t's
+    // identity field.
+
+    OE_TRACE_ERROR("modulus_size=%ld\n", modulus_size);
+    if (Sha256(modulus, modulus_size, signer) != 0)
+    {
+        OE_TRACE_ERROR("Sha256 failed\n");
+        goto done;
+    }
+
+    if (memcmp(signer, signer_id_buf, signer_id_buf_size) != 0)
+    {
+        OE_TRACE_ERROR("mrsigner is not equal!\n");
+        for (size_t i = 0; i < signer_id_buf_size; i++)
+        {
+            OE_TRACE_ERROR(
+                "0x%x - 0x%x\n", (uint8_t)signer[i], (uint8_t)signer_id_buf[i]);
+        }
+        goto done;
+    }
+    ret = true;
+done:
+    if (signer)
+        free(signer);
+
+    if (modulus != NULL)
+        free(modulus);
+
+    mbedtls_pk_free(&ctx);
```

Comment:
If signer is NULL, you will go to done without initing ctx. And then you are freeing it here before it was inited.

Reply:
Good catch. Addressed by moving the ctx initialization up